src/ppx/xr_component.cpp

```diff
@@ -0,0 +1,395 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if defined(PPX_BUILD_XR)
+#include "ppx/xr_component.h"
+#include "ppx/grfx/grfx_instance.h"
+#include <glm/gtc/type_ptr.hpp>
+
+namespace ppx {
+void XrComponent::InitializeBeforeGrfxDeviceInit(bool enableDebug)
+{
+    const char* graphicsAPIExtension = "";
+    // Extensions (Non-Optional)
+#if defined(PPX_D3D11)
+    graphicsAPIExtension = XR_KHR_D3D11_ENABLE_EXTENSION_NAME;
+#endif // defined(PPX_D3D11)
+
+#if defined(PPX_D3D12)
+    graphicsAPIExtension = XR_KHR_D3D12_ENABLE_EXTENSION_NAME;
+#endif // defined(PPX_D3D12)
+
+#if defined(PPX_VULKAN)
+    graphicsAPIExtension = XR_KHR_VULKAN_ENABLE2_EXTENSION_NAME;
+#endif // defined(PPX_VULKAN)
+
+    std::vector<const char*> xrInstanceExtensions;
+    xrInstanceExtensions.push_back(graphicsAPIExtension);
+    if (enableDebug) {
+        xrInstanceExtensions.push_back(XR_EXT_DEBUG_UTILS_EXTENSION_NAME);
+    }
+    // Verify extensions
+    uint32_t extCount = 0;
+    xrEnumerateInstanceExtensionProperties(nullptr, 0, &extCount, nullptr);
+    std::vector<XrExtensionProperties> xrExts(extCount, {XR_TYPE_EXTENSION_PROPERTIES});
+    xrEnumerateInstanceExtensionProperties(nullptr, extCount, &extCount, xrExts.data());
+    for (const auto& ext : xrInstanceExtensions) {
+        auto it = std::find_if(xrExts.begin(), xrExts.end(), [&](const XrExtensionProperties& e) { return strcmp(e.extensionName, ext) == 0; });
+        PPX_ASSERT_MSG((it != xrExts.end()), "The extension is not supported! Also make sure your OpenXR runtime is loaded properly!");
+    }
+
+    // Layers (Optional)
+    std::vector<const char*> xrRequestedInstanceLayers;
+    if (enableDebug) {
+        // This requires adding following lines in Environment in Debugging Settings
```

Comment:
This comment is VS specific. Make it more general by just saying "The following environment variables must be set:"

Reply:
fixed